### PR TITLE
feat(router): introduce server support for form actions

### LINKED ISSUE #935: Feature: Server Form Actions
### Which scope/s are relevant/related to the feature request?

trpc

### Information

Not sure if this would be a TRPC thing, or router thing.

It would be cool to have server actions. They are executed from a form action element. The goal is not just to submit to a form, but to all the form to have "progressive enhancements" so that they could still be submitted without JavaScript (if the internet is slow), and you get the same results as when you use JavaScript.

https://nextjs.org/docs/app/building-your-application/data-fetching/server-actions-and-mutations#behavior
https://kit.svelte.dev/docs/form-actions#progressive-enhancement
https://start.solidjs.com/core-concepts/actions
https://remix.run/docs/en/main/route/action
https://qwik.dev/docs/action/
https://form-actions-nuxt.pages.dev/get-started/introduction

### Describe any alternatives/workarounds you're currently using

Just manually send a form to an endpoint without progressive enhancements.  Just throwing this out there to have it on the backlog.

### I would be willing to submit a PR to fix this issue

- [ ] Yes
- [ ] No

### PR DESCRIPTION
**Pull request by brandonroberts**
## PR Checklist

Closes #935 

## What is the new behavior?

In `.server.ts` files, an `action` function can be defined to be used with forms to progressively enhance forms and build the foundation for handling form actions on the server.

```ts
// newsletter.server.ts
import {
  type PageServerAction,
  redirect,
  json,
  fail,
} from '@analogjs/router/server/actions';
import { readFormData } from 'h3';

export function load() {
  return {
    loaded: true,
  };
}

export async function action({ event }: PageServerAction) {
  const body = await readFormData(event);
  const email = body.get('email') as string;

  if (!email) {
    return fail(422, { email: 'Email is required' });
  }

  if (email.length < 10) {
    return redirect('/');
  }

  console.log({ email: body.get('email') });

  return json({ type: 'success' });
}
```

A follow-up PR will land to add a form directive to bridge the form to the form server action.

## Does this PR introduce a breaking change?

- [ ] Yes
- [x] No

### CODE REVIEW

File: apps/analog-app/src/app/pages/form-action.directive.ts [hunk 1]
```diff
@@ -0,0 +1,40 @@
...
+export class FormActionDirective {
+  action = input<string>('');
+  onSuccess = output<unknown>();
+  onError = output<unknown>();
+  state = output<'submitting' | 'error'>();
+  router = inject(Router);
+
+  submitted($event: { target: HTMLFormElement } & Event) {
+    $event.preventDefault();
+    const body = new FormData($event.target);
```
Comment (ajitzero) [inline apps/analog-app/src/app/pages/form-action.directive.ts:24]: I didn't know we could do that 🤦 I was using viewChild for this as I couldn't find this info within `$event` 🤷

File: apps/analog-app/src/app/pages/form-action.directive.ts [hunk 2]
```diff
@@ -0,0 +1,40 @@
...
+
+    fetch(this.action() || `/api/_analog/pages${window.location.pathname}`, {
+      method: $event.target.method,
+      body,
+    }).then((res) => {
```
Comment (ajitzero) [inline apps/analog-app/src/app/pages/form-action.directive.ts:29]: question: I'm not against `fetch`, but would it be possible to support/use HttpClient? Does the internal implementation even matter? I genuinely don't know.

File: apps/analog-app/src/app/pages/newsletter.server.ts [hunk 3]
```diff
@@ -0,0 +1,53 @@
...
+}
+
+export async function action({
+  event,
+}: PageServerAction): Promise<ActionResult> {
+  const body = await readFormData(event);
+  const email = body.get('email') as string;
+
+  if (!email) {
+    return fail(422, { errors: { email: 'Email is required' } });
```
Comment (ajitzero) [inline apps/analog-app/src/app/pages/newsletter.server.ts:43]: question: If JavaScript is disabled, what happens in this case? Can we set up something like `[...not-found].page.ts` but like `[...422].page.ts` to handle this?

Comment (ajitzero) [reply apps/analog-app/src/app/pages/newsletter.server.ts:43]: I tried handling it manually by adding a `format: 'json'` field if the call is made via Js:
- https://github.com/ajitzero/spartan-repro/blob/feat/form-action/analog-app/src/server/routes/v1/subscribe.ts
- https://github.com/ajitzero/spartan-repro/blob/feat/form-action/analog-app/src/app/pages/subscribe/(home).page.ts#L54

**Review by ajitzero — APPROVED** [review]

File: apps/analog-app/src/app/pages/form-action.directive.ts (hunk 2 above)
Comment (brandonroberts) [reply apps/analog-app/src/app/pages/form-action.directive.ts]: Its possible to use HttpClient, but I don't think it matters. fetch is more flexible out of the box in terms of determining what to do with the response

File: apps/analog-app/src/app/pages/newsletter.server.ts (hunk 3 above)
Comment (brandonroberts) [reply apps/analog-app/src/app/pages/newsletter.server.ts]: This is still TBD. Ideally, if JS is off, you come back to this page with errors displayed as part of the rendered HTML

**Comment by jdgamble555** [comment]
whahoo!

File: apps/analog-app/src/app/pages/search.server.ts [hunk 4]
```diff
@@ -0,0 +1,35 @@
...
+  const query = getQuery(event);
+  console.log('loaded search', query['search']);
+
+  return {
+    loaded: true,
+    searchTerm: `${query['search']}`,
+  };
+}
+
+export async function action({ event }: PageServerAction) {
```
Comment (nartc) [inline apps/analog-app/src/app/pages/search.server.ts:20]: nit: this is copy-pasta so not needed? If this is for demo purposes, can we change the body to something related to the search page instead?

**Review by nartc — APPROVED** [review]
Approved with one small nit

File: apps/analog-app/src/app/pages/search.server.ts (hunk 4 above)
Comment (brandonroberts) [reply apps/analog-app/src/app/pages/search.server.ts:20]: Oh yea, copy-pasta. Will remove

File: apps/analog-app/src/app/pages/form-action.directive.ts [hunk 5]
```diff
@@ -0,0 +1,78 @@
...
+              this.router.navigate([redirectUrl]);
+            } else if (res.headers.get('Content-type') === 'application/json') {
+              res.json().then((result) => {
+                this.onSuccess.emit(result);
+                this.state.emit('success');
+              });
+            } else {
+              res.text().then((result) => {
+                this.onSuccess.emit(result);
+                this.onSuccess.emit('success');
```
[the remark this answers is not in the document]
Comment (brandonroberts) [reply apps/analog-app/src/app/pages/form-action.directive.ts:59]: Yep

**Review by Jevgenijs92 — APPROVED** [review]

**Pull request merged**